Adds section to describe intended usage of `pipeline.buffer.type` (#16083)
elastic/logstash: docs/static/config-details.asciidoc

Keep these memory requirements in mind as you calculate your ideal memory allocation.

> [[reducing-off-heap-usage]]

Reviewer: I wonder if we should better highlight this somehow. The use of this setting completely changes the way Logstash works. Also, it would give us the ability to better understand Logstash when troubleshooting.

However, the direct result of this is that will completely impact/affect JVM Heap sizing. So I propose we:

1. Bring some more visibility into this setting, maybe say this will be the default in future versions of Logstash.
2. Tell users that this is a recommended setting to be set?
3. And, in the memory size/performance tuning sections to have references to this setting being turned on??

Any thoughts?

@jsvd @flexitrev

Java Heap instead of direct memory. In this way, in case of out-of-memory conditions, a better report of eventual leak
could be investigated.
Moving the allocation of IO buffers to Java heap generally doesn't impact the performance in terms of requests per second,
because every event's payload has however to be converted from raw bytes in direct memory space to Logstash events,
which are Java objects that live in the heap.
However, the Java heap sizing requirements could be impacted by this change, because it has to allocate space that previously resided
on the direct zone, outside heap. In such case you could increment the max size of Java heap by the amount of the direct space,
but don't reduce too mush the direct memory size because there could be some allocation that still happens there.

NOTE: When `pipeline.buffer.type` is valued to `heap` consider to increment the Java heap by the amount of
previously reserved for direct space. At the same time do not decrease too much of the direct space, you could try to
set to the half of the previous size.

NOTE: `pipeline.buffer.type` is not intended to provide better throughput of events per second but to have better insights

> when out-of-memory in direct space happens. However, it should not impact the overall performance.

Author: @karenzone after integrating the suggestion something breaks, but I all the asciidoc tags and syntax to me seems correct. Do you have any hint why it fail?

Keep these memory requirements in mind as you calculate your ideal memory allocation.

[[reducing-off-heap-usage]]

> ===== Troubleshooting out of memory errors in off-heap space: `pipeline.buffer.type`

Reviewer 2: ```suggestion
===== Upcoming changes to Buffer Allocation and Troubleshooting Out of Memory errors
```

> Some plugins, such as {agent}, {beats}, TCP, and HTTP inputs, use direct memory to get

Reviewer 2: ```suggestion
Plugins such as {agent}, {beats}, TCP, and HTTP inputs, use direct memory to get
```

> Plugins such as {agent}, {beats}, TCP, and HTTP inputs, use direct memory to get
> better performance when interacting with the network stack. 

Reviewer 2: ```suggestion
Plugins such as {agent}, {beats}, TCP, and HTTP inputs, currently default to using direct memory as it tends
to provide better performance, especially when interacting with the network stack. 
```

> Under certain circumstances of heavy load, the direct memory space can be exhausted and lead to Out of Memory (OOM) errors in off-heap space.

Reviewer 2: ```suggestion
Under heavy load, namely large number of connections and large messages, the direct memory space can be exhausted and lead to Out of Memory (OOM) errors in off-heap space.
```

> {ls} provides a `pipeline.buffer.type` setting in <<logstash-settings-file>> that can help you investigate and troubleshoot Out of Memory errors. 
> This setting lets you control where to allocate memory buffers for plugins that use them. 
> It is set to `direct` by default, but you can change it to `heap` to use Java heap space instead.
> When set to `heap`, all IO buffer allocations used by plugins are configured to **prefer** the
> Java Heap instead of direct memory. 

Reviewer 2: ```suggestion
An off-heap OOM is difficult to debug, so {ls} provides a `pipeline.buffer.type` setting in <<logstash-settings-file>> that lets you control where to allocate memory buffers for plugins that use them. 
Currently it is set to `direct` by default, but you can change it to `heap` to use Java heap space instead, which will be become the default in the future.
When set to `heap`, buffer allocations used by plugins are configured to **prefer** the
Java Heap instead of direct memory, as direct memory allocations may still be necessary depending on the plugin.
```

> In the case of an out-of-memory condition, you have a better report for investigating memory leaks.

Reviewer 2: ```suggestion
When set to "heap", in the event of an out-of-memory, Logstash will produce a heap dump to facilitate debugging.
```

Moving the allocation of IO buffers to Java heap generally doesn't impact the performance in terms of requests per second.
However, every event's payload has to be converted from raw bytes in direct memory space to {ls} events,
which are Java objects that live in the heap.

> Note that the Java heap sizing requirements could be impacted by this change because
> it has to allocate space that previously resided on the direct memory. 

Reviewer 2: ```suggestion
Note that the Java heap sizing requirements will be impacted by this change since
allocations that previously resided on the direct memory will use heap instead. 
```

> Moving the allocation of IO buffers to Java heap generally doesn't impact the performance in terms of requests per second.
> However, every event's payload has to be converted from raw bytes in direct memory space to {ls} events,
> which are Java objects that live in the heap.
> 
> Note that the Java heap sizing requirements will be impacted by this change since
> allocations that previously resided on the direct memory will use heap instead. 

Reviewer 2: ```suggestion
It is important to note that the Java heap sizing requirements will be impacted by this change since
allocations that previously resided on the direct memory will use heap instead. 

Performance-wise there shouldn't be a noticeable impact, since while direct memory IO is faster, Logstash Event objects produced by these plugins end up being allocated on the Java Heap, incurring the cost of copying from direct memory to heap memory regardless of the setting.
```

it has to allocate space that previously resided on the direct memory. 

[NOTE] 
--
* When you set `pipeline.buffer.type` to `heap`, consider incrementing the Java heap by the 
amount of memory that had been reserved for direct space. 

> 

Reviewer 3: ```suggestion
--

```

Keep these memory requirements in mind as you calculate your ideal memory allocation.

[[reducing-off-heap-usage]]

> [[oom-off-heap-usage]]

Reviewer 3: Use one or the other, but not both.  Your choice.  :-)

> [[oom-off-heap-usage]]

Reviewer 3: We have two anchors, so please delete this one.
